ui: fix jobs page crash

vladlos opened the pull request:
while flipping pages on jobs table user could encount page crash. in addition
to that there was a visible delay beetwen data changes that make no sence
since pagination itself is pure ui.
there were a compbination of issues that cause such behavior, this pr contains:
 - on my local setup I got jobs with empty `statement` and `description` for such
case I added fallback to `type` value, to at least leave some value in that column as
it works also as a link and so there should be some clickable value
 - fix for unneeded component rerenders
 - after removing extra rerenders another issue with table component become
visible, wrong use of reselect library, fixed
 - combination last 2 changes fixed issue with slow data refresh
 - in addition fixed font color on `description` value tooltips, it was wrong and barely
visible

Resolves: #62905

Release note (ui): fix jobs page crash while using pagination, improve performance

![Screenshot 2021-05-24 at 16 45 12](https://user-images.githubusercontent.com/12850886/119377207-8754e880-bcc5-11eb-85bc-4d7d6163a1e4.png)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65618)
<!-- Reviewable:end -->

koorosh reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65618)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster and @nathanstilwell)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaYCe8UCF7nx_mAXPBQ:-MaYCe8UCF7nx_mAXPBR:b-mp5lr4) (no related file):*
@elkmaster , I tried to debug the main issue with crashed Jobs page and found that the only fix required is to apply following change in `sortedAndPaginated` selector:
```typescript
  sortedAndPaginated = createSelector(
    (props: SortedTableProps<T>) => props.data,
    (props: SortedTableProps<T>) => props.sortSetting,
    (props: SortedTableProps<T>) => props.columns,
    (props: SortedTableProps<T>) => props.pagination, // <-- add this line 
    (
      data: T[],
      sortSetting: SortSetting,
      columns: ColumnDescriptor<T>[],
      _pagination, // <-- it's not used but is registered as a dependency for selector to trigger re-computation
    ): T[] => {
...
```

What I see, the problem occurs when `sortedAndPaginated` selector memoizes an old pagination value (especially for the last page which has less records that other pages) and then selector doesn't recompute records (when another page is selected) and returns incorrect number of records for the last page.

---


<!-- Sent from Reviewable.io -->

vladlos reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65618)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh and @nathanstilwell)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaYCe8UCF7nx_mAXPBQ:-MacrWe36QvZSwdie5Bw:bxhmkwb) (no related file):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

@elkmaster , I tried to debug the main issue with crashed Jobs page and found that the only fix required is to apply following change in `sortedAndPaginated` selector:
```typescript
  sortedAndPaginated = createSelector(
    (props: SortedTableProps<T>) => props.data,
    (props: SortedTableProps<T>) => props.sortSetting,
    (props: SortedTableProps<T>) => props.columns,
    (props: SortedTableProps<T>) => props.pagination, // <-- add this line 
    (
      data: T[],
      sortSetting: SortSetting,
      columns: ColumnDescriptor<T>[],
      _pagination, // <-- it's not used but is registered as a dependency for selector to trigger re-computation
    ): T[] => {
...
```

What I see, the problem occurs when `sortedAndPaginated` selector memoizes an old pagination value (especially for the last page which has less records that other pages) and then selector doesn't recompute records and returns incorrect number of records for the last page.


</blockquote></details>

cleaned up extensive null checks, seems they are not really needed after original fix for selector

---


<!-- Sent from Reviewable.io -->

nathanstilwell reviewed (approved):
LGTM

koorosh reviewed (changes requested):
> Reviewed 6 of 6 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65618)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaeZ6-T6uv_kHuH5zkq:-MaeZ6-U7BGAV2Bp_MBi:b-fqa2lr) (no related file):*
Still have concerns about checks whether component has to be updated or not, please see comments below.

---
___
*[pkg/ui/src/views/jobs/index.tsx, line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaeGj7U5x0ei3VbaYX9:-MaeGj7U5x0ei3VbaYXA:b-2iux7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca597083a4e1b6d6e00b49f6cf3f7ef9f4531769/pkg/ui/src/views/jobs/index.tsx#L117)):*
> ```XML
>   shouldComponentUpdate(nextProps: JobsTableProps) {
>     const { sort, status, show, type, jobs } = this.props;
>     if (
> ```

This condition includes most of props (except functions). I suppose it has the same effect as default props checks for component updates. 
Does it really affects component updates rate?
___
*[pkg/ui/src/views/jobs/jobTable.tsx, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaeDxnp6tEaIQ7CYqSp:-MaeDxnp6tEaIQ7CYqSq:b-qk9xkf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca597083a4e1b6d6e00b49f6cf3f7ef9f4531769/pkg/ui/src/views/jobs/jobTable.tsx#L44)):*
> ```XML
>     className: "cl-table__col-query-text",
>     cell: (job) => <JobDescriptionCell job={job} />,
>     sort: (job) => job.description || job.type,
> ```

`JobDesscriptionCell` component uses following condition for Description field.
```{job.statement || job.description || job.type}```
Probably the same condition has to be applied for sorting as well?
___
*[pkg/ui/src/views/jobs/jobTable.tsx, line 171 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaeF2q-4dXS7srhYBiL:-MaeF2q-4dXS7srhYBiM:bkjhxbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca597083a4e1b6d6e00b49f6cf3f7ef9f4531769/pkg/ui/src/views/jobs/jobTable.tsx#L171)):*
> ```XML
>     const { sort, jobs, pageSize, current, isUsedFilter } = this.props;
> 
>     if (
> ```

@elkmaster , what is the purpose for limiting component updates with these changes?
As I understood, current changes include all props but does ignore component's `state` change that includes pagination.
I assume that this component can be updated with every change, and there is no need for custom `shouldComponentUpdate` function.
___
*[pkg/ui/src/views/shared/components/sortedtable/index.tsx, line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaeCSMP-ZcdLRkJwbAw:-MaeCSMP-ZcdLRkJwbAx:bb9gcw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca597083a4e1b6d6e00b49f6cf3f7ef9f4531769/pkg/ui/src/views/shared/components/sortedtable/index.tsx#L142)):*
> ```XML
>       pagination: ISortedTablePagination,
>     ): T[] => {
>       const paginatedData = (sortData?: T[]) => {
> ```

What is the reason for moving function declaration inside of selector? Would it be okay to define it as a pure function?


<!-- Sent from Reviewable.io -->

vladlos committed:
ui: fix jobs page crash

while flipping pages on jobs table user could encount page crash. in addition
to that there was a visible delay beetwen data changes that make no sence
since pagination itself is pure ui.
there were a compbination of issues that cause such behavior, this pr contains:
- on my local setup I got jobs with empty `statement` and `description` for such
case I added fallback to `type` value, to at least leave some value in that column as
it works also as a link and so there should be some clickable value
- fix for unneeded component rerenders
- after removing extra rerenders another issue with table component become
visible, wrong use of reselect library, fixed
- combination last 2 changes fixed issue with slow data refresh
- in addition fixed font color on `description` value tooltips, it was wrong and barely
visible

Resolves: cockroachdb#62905

Release note (ui): fix jobs page crash while using pagination, improve performance

vladlos reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65618)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaeZ6-T6uv_kHuH5zkq:-Mahl5rT8bMKC0xNdAsO:bbkjvj5) (no related file):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Still have concerns about checks whether component has to be updated or not, please see comments below.
</blockquote></details>

done, see comments below.

---
___
*[pkg/ui/src/views/jobs/index.tsx, line 117 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaeGj7U5x0ei3VbaYX9:-MahkhZY5xWK_rDqc55l:br0plj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca597083a4e1b6d6e00b49f6cf3f7ef9f4531769/pkg/ui/src/views/jobs/index.tsx#L117)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

This condition includes most of props (except functions). I suppose it has the same effect as default props checks for component updates. 
Does it really affects component updates rate?
</blockquote></details>

you are right, to avoid refresh loop componentDidUpdate changes are enough
___
*[pkg/ui/src/views/jobs/jobTable.tsx, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaeDxnp6tEaIQ7CYqSp:-MahkqrnDFJ18iSgLHUr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca597083a4e1b6d6e00b49f6cf3f7ef9f4531769/pkg/ui/src/views/jobs/jobTable.tsx#L44)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

`JobDesscriptionCell` component uses following condition for Description field.
```{job.statement || job.description || job.type}```
Probably the same condition has to be applied for sorting as well?

</blockquote></details>

Done.
___
*[pkg/ui/src/views/shared/components/sortedtable/index.tsx, line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-MaeCSMP-ZcdLRkJwbAw:-Mahks_v2WRCVVYqY4Fk:bdaes3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca597083a4e1b6d6e00b49f6cf3f7ef9f4531769/pkg/ui/src/views/shared/components/sortedtable/index.tsx#L142)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

What is the reason for moving function declaration inside of selector? Would it be okay to define it as a pure function?
</blockquote></details>

it is there just because it should be used only inside selector, but is it ok to move it out for readability reasons


<!-- Sent from Reviewable.io -->

koorosh reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65618#-:-Mahsfw81uAcOfEher9w:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @elkmaster)



<!-- Sent from Reviewable.io -->

elkmaster commented:
bors r+
